idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,248 @@
+            - results.theta(dict)                   : Estimated parameters
+            - results.theta_names(dict)             : Names of parameters
+            - results.cov(numpy.ndarray)            : Covariance of theta
+            - results.gradient_f(numpy.ndarray)     : Gradient vector of the objective function with respect to the (decision variables, parameters) at the optimal solution
+            - results.gradient_c(scipy.sparse.coo.coo_matrix)     : Gradient vector of the constraints with respect to the (decision variables, parameters) at the optimal solution. Each row contains [column number, row number, and value], colum order follows variable order in col. If no constraint exists, return []. Note: Python sparse matrix index starts from 0.  To match with k_aug sparse form, need to add 1 for each row and col index.
+            - results.dsdp(numpy.ndarray)           : Gradient vector of the (decision variables, parameters) with respect to paramerters (=theta_name). number of rows = len(theta_name), number of columns= len(col)
+            - results.propagation_c(numpy.ndarray)  : Error propagation in the constraints (only constraints have theta_name)  e.g) [[constraint iumber1, error1],[constraint number2, error2]]             
+            - results.propagation_f(numpy.float64)  : Error propagation in the objective function, df/dp*cov_p*df/dp + (df/dx*dx/dp)*cov_p*(df/dx*dx/dp)
+            - results.col(list)                     : List of variable names  
+    
+    Raises:
+        TypeError:  When tee entry is not Boolean
+        TypeError:  When diagnostic_mode entry is not Boolean
+        TypeError:  When solver_options entry is not None and a Dictionary
+        Warnings:   When an element of theta_names includes a space
+    
+    """
+
+    if not isinstance(tee, bool):
+        raise TypeError('tee  must be boolean.')
+    if not isinstance(diagnostic_mode, bool):
+        raise TypeError('diagnostic_mode  must be boolean.')    
+    if not solver_options==None:
+        if not isinstance(solver_options, dict):
+            raise TypeError('solver_options must be dictionary.')
+    # Remove all "'" and " " in theta_names
+    theta_names, var_dic,variable_clean = clean_variable_name(theta_names)
+    parmest_class = parmest.Estimator(model_function, data, theta_names, obj_function,
+                 tee, diagnostic_mode, solver_options)
+    obj, theta,cov = parmest_class.theta_est(calc_cov=True)
+    # Convert theta keys to the original name 
+    # Revert theta_names to be original
+    if variable_clean:
+        theta_out = {}
+        for i in var_dic.keys():
+            theta_out[var_dic[i]] = theta[i]
+        theta_names = [var_dic[v] for v in theta_names] 
+    else:
+        theta_out = theta
+    propagate_results  =  propagate_uncertainty(model_uncertain, theta, cov, theta_names, tee)
+   
+    if len(propagate_results.gradient_c)>0:
```

Comment:
Please add a comment. What is going on here?

Reply:
My recommendation is for `propagate_results` to contain a sparse matrix. This way, you do not need to create one here. It makes the code more modular.

Reply:
@adowling2 
For both `propagation_f` and `propagation_c`, I think we can calculate 
- `propagation_f = df/dp*cov_p*df/dp + (df/dx*dx/dp)*cov_p*(df/dx*dx/dp)= df/ds * ds/dp * cov * df/ds * ds/dp`. So, I don't need to split the variables into decision and parameters. I changed codes like the below

        #step 1. df/ds*ds/dp**
        fssp = np.matmul(np.reshape(gradient_f,(1,len(gradient_f))),np.reshape(dsdp_new (len(dsdp_new),len(theta_names))))
        # step 2. (df/ds*ds/dp)*cov*(df/ds*ds/dp)
        propagation_f = np.sum(np.multiply(np.matmul(fssp,cov),fssp))
- `propagation_c = dc/dp*cov_p*dc/dp + (dc/dx*dx/dp)*cov_p*(dc/dx*dx/dp)= dc/ds * ds/dp * cov * dc/ds * ds/dp`

        #step 1. dc/ds*ds/dp**
        cssp = np.matmul(gradient_c.toarray(), dsdp_new)
        # step 2. (df/ds*ds/dp)*cov*(df/ds*ds/dp)
        propagation_c = np.sum(np.matmul(cssp,cov)*cssp,axis=1)